armi/nuclearDataIO/rtflux.py

```diff
@@ -24,180 +41,172 @@
-        # Also read index pointers to incoming partial currents on outer reactor surface
-        # (these don't belong to any assembly)
-        # Incoming partial currents are non-zero due to flux extrapolation
-        ng = self.fc["ngroup"]  # number of energy groups
-        imax = self.fc[
-            "ninti"
-        ]  # number of triangular mesh cells in "i" direction (rhombus or rectangle cells)
-        jmax = self.fc[
-            "nintj"
-        ]  # number of triangular mesh cells in "j" direction (rhombus or rectangle cells)
-        zmax = self.fc[
-            "nintk"
-        ]  # number of axial nodes (same for each assembly in DIF3D)
+    def __init__(self):
+        # Need Metadata subclass for default keys
+        self.metadata = nuclearFileMetadata._Metadata()
 
-        self.triangleFluxes = numpy.zeros((imax, jmax, zmax, ng))
+        self.groupFluxes = None
 
-        for g in range(ng):  # loop through energy groups
 
-            gEff = self.getEnergyGroupIndex(g)
+class RtfluxStream(cccc.Stream):
+    """
+    Stream for reading/writing a RTFLUX or ATFLUX file.
 
-            for z in range(zmax):  # loop through axial nodes
-                self.triangleFluxes[
-                    :, :, z, gEff
-                ] = self.readTriangleFluxes()  # read fluxes on this i-j plane
+    Parameters
+    ----------
+    flux : RtfluxData
+        Data structure
+    fileName: str
+        path to RTFLUX file
+    fileMode: str
+        string indicating if ``fileName`` is being read or written, and
+        in ascii or binary format
 
-        self.f.close()
+    """
 
-    def read1D(self):
-        r"""
-        Read parameters from the RTFLUX/ATFLUX 1D block (file control).
+    def __init__(self, flux: RtfluxData, fileName: str, fileMode: str):
+        cccc.Stream.__init__(self, fileName, fileMode)
+        self._flux = flux
+        self._metadata = self._getFileMetadata()
+
+    def _getFileMetadata(self):
+        return self._flux.metadata
+
+    @classmethod
+    def _read(cls, fileName: str, fileMode: str) -> RtfluxData:
+        flux = RtfluxData()
+        return cls._readWrite(
+            flux,
+            fileName,
+            fileMode,
+        )
 
-        This contains a bunch of single-number integer or double values
-        necessary to read all data from the other records of RTFLUX/ATFLUX.
-        See the comments following each quantity in the code.
+    # pylint: disable=arguments-differ
+    @classmethod
+    def _write(cls, flux: RtfluxData, fileName: str, fileMode: str):
+        return cls._readWrite(flux, fileName, fileMode)
 
-        See Also
-        --------
-        rtflux.RTFLUX.__init__
-        nuclearDataIO.ISOTXS.read1D
-        nuclearDataIO.SPECTR.read1D
+    @classmethod
+    def _readWrite(cls, flux: RtfluxData, fileName: str, fileMode: str) -> RtfluxData:
+        with RtfluxStream(flux, fileName, fileMode) as rw:
+            rw.readWrite()
+        return flux
```

Comment:
Pretty squirrely, but 🤷‍♂️. How would I know how I am supposed to use this class? Seems like all of the client-facing stuff is actually on the `cccc.Stream` class? An example in the class docstring would be nice. Or an example in the `RtFluxData` of how to read one of those from one of these would be useful

Reply:
I think we might be shooting ourselves in the foot with the DRY-cannon with the structure of the cccc classes. Not worth fixing in this PR, but something to think about.

Reply:
I'm sure we can reduce the boilerplate but I am just copy/pasting it rather than DRYing it. I could have yet another parent class for all these CCCC rw-ers that specializes their data container. 

I used to think these were really complex but once you see the light it's quite glorious to be able to define the file structure once and get free reading AND writing!

Reply:
Sure, but that's an *implementer's* paradise, not a client's. Especially in the context of a file format reader/writer, these classes will be the subject of someone trying to figure out how to use them considerably more often than someone trying to create/modify one. Classic "you write code once, but read it many times" situation. Someone that thinks to themselves "ooh look, an RTFLUX reader! let's read me an RTFLUX," then proceeding to try to figure out how will probably struggle to, because the methods that they actually want to use are defined two classes away.